etna/transforms/encoders/mean_encoder.py

```diff
@@ -0,0 +1,309 @@
+                        feature.loc[nan_feature_index] = expanding_mean.loc[nan_feature_index]
+                    intersected_df.loc[:, self.idx[segment, self.out_column]] = feature.values
+
+            else:
+                flatten = TSDataset.to_flatten(intersected_df)
+                flatten = flatten.sort_values(["timestamp", "segment"])
+                running_mean = self._count_macro_running_mean(flatten, n_segments)
+
+                temp = pd.DataFrame(index=flatten.index, columns=["cumsum", "cumcount"], dtype=float)
+
+                timestamps = intersected_df.index
+                categories = pd.unique(df.loc[:, self.idx[:, self.in_column]].values.ravel())
+
+                cumstats = pd.DataFrame(data={"sum": 0, "count": 0, self.in_column: categories})
+                start_index = np.arange(0, len(timestamps) * n_segments, len(timestamps))
+                for timestamp in timestamps:
+                    timestamp_df = TSDataset.to_flatten(intersected_df.loc[timestamp:timestamp, self.idx[:, :]])
+
+                    cumsum_dict = dict(cumstats[[self.in_column, "sum"]].values)
+                    cumcount_dict = dict(cumstats[[self.in_column, "count"]].values)
+
+                    temp.loc[start_index, "cumsum"] = flatten.loc[start_index, self.in_column].map(cumsum_dict)
+                    temp.loc[start_index, "cumcount"] = flatten.loc[start_index, self.in_column].map(cumcount_dict)
+
+                    stats = (
+                        timestamp_df["target"]
+                        .groupby(timestamp_df[self.in_column], dropna=False)
+                        .agg(["count", "sum"])
+                        .reset_index()
+                    )
+                    cumstats = pd.concat([cumstats, stats]).groupby(self.in_column, as_index=False, dropna=False).sum()
+                    start_index += 1
+
+                feature = (temp["cumsum"] + running_mean * self.smoothing) / (temp["cumcount"] + self.smoothing)
+                if self.handle_missing is MissingMode.global_mean:
+                    nan_feature_index = flatten[flatten[self.in_column].isnull()].index
+                    feature.loc[nan_feature_index] = running_mean.loc[nan_feature_index]
+
+                feature = pd.DataFrame(
+                    feature.values.reshape(len(timestamps), n_segments),
+                    columns=pd.MultiIndex.from_product([segments, [self.out_column]]),
+                    index=intersected_df.index,
+                )
+                intersected_df = pd.concat([intersected_df, feature], axis=1)
+
+        if len(future_df) > 0:
+            n_timestamps = len(future_df.index)
+            if self.mode is EncoderMode.per_segment:
+                for segment in segments:
+                    segment_df = TSDataset.to_flatten(future_df.loc[:, self.idx[segment, :]])
+                    feature = segment_df[self.in_column].map(self._global_means_category[segment])
```

Comment:
Before this line you should add a casting like:
```
self._global_means_category = cast(..., self._global_means_category)
self._global_means = cast(Dict[str, float], self._global_means)
```
It will satisfy mypy.